### PR TITLE
Support for tiled decompositions in PPPM

### PR DESCRIPTION
**Pull request by sjplimp**
**Summary**

Add support to PPPM for load-balanced decompositions produced by balance and fix balance rcb,
i.e. tiled decompositions.  The majority of changes for this are in the GridComm class which now stores a regular or tiled decomposition of the FFT grids used by PPPM and does forward and reverse communication of the charge and efield values on those grids.  This change should support both orthogonal and triclinic domains and all the options that the PPPM class supports.  By the time it is ready to merge all variant PPPM classes, including accelerator versions, should be supported.

**Related Issues**

N/A

**Author(s)**

Steve

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Non-tiled decompositions should work as before.

**Implementation Notes**

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by sjplimp** [comment]
@akohlmey @stanmoore1 As we discussed this PR initially has new PPPM2 and GridComm2 classes with the added support for tiling.  The original PPPM and GridComm classes are unchanged.

**Comment by sjplimp** [comment]
@akohlmey Generalized this so the GridComm class can be used by fixes or computes.  Similar to how the Comm methods for forward/reverse comm can be called from pair, fix, compute styles.  Maybe this would be better done with a single templated forward_comm() method in GridComm2 which was templated on a ptr to a KSpace, Fix, Compute instance ?

**Comment by akohlmey** [comment]
@sjplimp ran this with the unittest test programs and found no difference between when switching to comm_style tiled, but those are all serial tests, so I am not sure how relevant those are. What worries me a little bit is that changing from orthogonal to triclinic changes the forces even though all the tilt factors are 0. This is the same for pppm an pppm2 though.

**Comment by stanmoore1** [comment]
> What worries me a little bit is that changing from orthogonal to triclinic changes the forces even though all the tilt factors are 0. This is the same for pppm an pppm2 though.

Are you manually constraining the grid? I believe the logic for choosing the grid size is a little different for triclinic vs not.

**Comment by akohlmey** [comment]
> > What worries me a little bit is that changing from orthogonal to triclinic changes the forces even though all the tilt factors are 0. This is the same for pppm an pppm2 though.
> 
> Are you manually constraining the grid? I believe the logic for choosing the grid size is a little different for triclinic vs not.

nope. there are the test inputs and lists of reference forces/energies/stresses: https://github.com/lammps/lammps/blob/master/unittest/force-styles/tests/kspace-pppm.yaml and https://github.com/lammps/lammps/blob/master/unittest/force-styles/tests/kspace-pppm_tri.yaml

all kspace related commands are in the "post_commands" section. The tester feeds input fragments, files and settings from these files into a LAMMPS instance and then compares the forces, stresses and energies.

**Comment by stanmoore1** [comment]
You will get slightly different forces if triclinic picks a 34x34x34 grid while orthogonal picks a 32x32x32 grid.

**Comment by sjplimp** [comment]
@akohlmey @stanmoore1 Did this difference between orthogonal and triclinic for an orthog geometry get resolved?  This was for both the original and this new PPPM2, so not an issue caused by this PR, correct?  I am about ready to convert this PPPM2 to PPPM and propagate the changes to other files.  Any reason to still wait on doing that?

**Comment by sjplimp** [comment]
@akohlmey @stanmoore1  I believe I have made all the needed mods to all the PPPM variants, except Kokkos. I also removed the trial PPPM2 and GridComm2 classes.  So now the changes are in the mainstream PPPM and GridComm class.  I think more changes are needed in GridCommKokkos than in PPPMKokkos.  MSM is more complicated.  I think we should disable tiled with MSM for now, but it will take some changes to MSM to make it work with the new GridComm API.  If you can test all the PPPM variants, w/out MSM in a good state, please do.

**Comment by sjplimp** [comment]
@akohlmey @stanmoore1 ok, with Stan's advice I made MSM compatible with the new GridComm class.  So everything but PPPM/Kokkos is now in place.  Please test everything you can think of.

**Comment by akohlmey** [comment]
> @akohlmey @stanmoore1 ok, with Stan's advice I made MSM compatible with the new GridComm class. So everything but PPPM/Kokkos is now in place. Please test everything you can think of.

@sjplimp While I look into this, I need to you to look at PR #2284

**Comment by akohlmey** [comment]
@sjplimp this branch current fails to compile kspace style msm/cg/omp.
Kspace style pppm/stagger produces incorrect forces.
It also was erroring out in the unit test for msm with a ` double free or corruption (!prev)` message after `MSM initialization`
and for kspace style msm/cg it has a segmentation fault at the same location. 
```
[==========] Running 6 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 6 tests from PairStyle
[ RUN      ] PairStyle.plain

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff752d7c0 in free () from /lib64/libc.so.6
Missing separate debuginfos, use: dnf debuginfo-install brotli-1.0.7-6.fc31.x86_64 cyrus-sasl-lib-2.1.27-3.fc31.x86_64 fftw-libs-double-3.3.8-6.fc31.x86_64 hwloc-libs-2.0.4-1.fc31.x86_64 keyutils-libs-1.6-3.fc31.x86_64 krb5-libs-1.17-46.fc31.x86_64 libcom_err-1.45.5-1.fc31.x86_64 libcurl-7.66.0-2.fc31.x86_64 libgcc-9.3.1-2.fc31.x86_64 libgfortran-9.3.1-2.fc31.x86_64 libgomp-9.3.1-2.fc31.x86_64 libidn2-2.3.0-1.fc31.x86_64 libjpeg-turbo-2.0.2-5.fc31.x86_64 libnghttp2-1.41.0-1.fc31.x86_64 libpng-1.6.37-2.fc31.x86_64 libpsl-0.21.0-2.fc31.x86_64 libquadmath-9.3.1-2.fc31.x86_64 libselinux-2.9-5.fc31.x86_64 libssh-0.9.4-3.fc31.x86_64 libstdc++-9.3.1-2.fc31.x86_64 libtool-ltdl-2.4.6-31.fc31.x86_64 libxcrypt-4.4.16-3.fc31.x86_64 libyaml-0.2.2-2.fc31.x86_64 mpich-3.3.2-1.fc31.x86_64 openssl-libs-1.1.1g-1.fc31.x86_64 zlib-1.2.11-20.fc31.x86_64
(gdb) up
#1  0x0000000000b3a5dc in LAMMPS_NS::MSM::deallocate_levels (this=this@entry=0x257de80) at /home/akohlmey/compile/lammps/src/KSPACE/msm.cpp:858
858	  delete [] gc_buf1;
(gdb) 
#2  0x0000000000b4b91a in LAMMPS_NS::MSM::set_grid_global (this=this@entry=0x257de80) at /home/akohlmey/compile/lammps/src/KSPACE/msm.cpp:1120
1120	  deallocate_levels();
```
In both cases initializing the affected pointers to NULL in the constructor resolved those issues.  This is done in commit bd7917919bd442883f5717f5f0300ed61c925987 after that the unittest result is:
```
The following tests passed:
	KSpaceStyle:ewald
	KSpaceStyle:ewald_disp
	KSpaceStyle:ewald_nozforce
	KSpaceStyle:ewald_slab
	KSpaceStyle:msm
	KSpaceStyle:msm_cg
	KSpaceStyle:pppm
	KSpaceStyle:pppm_ad
	KSpaceStyle:pppm_cg
	KSpaceStyle:pppm_cg_ad
	KSpaceStyle:pppm_disp
	KSpaceStyle:pppm_disp_ad
	KSpaceStyle:pppm_disp_tip4p
	KSpaceStyle:pppm_nozforce
	KSpaceStyle:pppm_slab
	KSpaceStyle:pppm_tip4p
	KSpaceStyle:pppm_tip4p_ad
	KSpaceStyle:pppm_tip4p_nozforce
	KSpaceStyle:pppm_tip4p_slab
	KSpaceStyle:pppm_tri

95% tests passed, 1 tests failed out of 21

Total Test time (real) =  26.38 sec

The following tests FAILED:
	196 - KSpaceStyle:pppm_stagger (Failed)
```
So only this and the failed compilation of msm/cg/omp remains:
```
[  8%] Building CXX object CMakeFiles/lammps.dir/home/akohlmey/compile/lammps/src/USER-OMP/msm_cg_omp.cpp.o
/home/akohlmey/compile/lammps/src/USER-OMP/msm_cg_omp.cpp: In member function ‘virtual void LAMMPS_NS::MSMCGOMP::compute(int, int)’:
/home/akohlmey/compile/lammps/src/USER-OMP/msm_cg_omp.cpp:100:5: error: ‘cg_peratom_all’ was not declared in this scope
  100 |     cg_peratom_all->ghost_notify();
      |     ^~~~~~~~~~~~~~
/home/akohlmey/compile/lammps/src/USER-OMP/msm_cg_omp.cpp:104:7: error: ‘cg_peratom’ was not declared in this scope; did you mean ‘direct_peratom’?
  104 |       cg_peratom[n]->ghost_notify();
      |       ^~~~~~~~~~
      |       direct_peratom
/home/akohlmey/compile/lammps/src/USER-OMP/msm_cg_omp.cpp:178:3: error: ‘cg_all’ was not declared in this scope; did you mean ‘gcall’?
  178 |   cg_all->reverse_comm(this,REVERSE_RHO);
      |   ^~~~~~
      |   gcall
/home/akohlmey/compile/lammps/src/USER-OMP/msm_cg_omp.cpp:186:5: error: ‘cg’ was not declared in this scope; did you mean ‘gc’?
  186 |     cg[n]->forward_comm(this,FORWARD_RHO);
      |     ^~
      |     gc
/home/akohlmey/compile/lammps/src/USER-OMP/msm_cg_omp.cpp:199:7: error: ‘cg’ was not declared in this scope; did you mean ‘gc’?
  199 |       cg[levels-1]->forward_comm(this,FORWARD_RHO);
      |       ^~
      |       gc
/home/akohlmey/compile/lammps/src/USER-OMP/msm_cg_omp.cpp:203:9: error: ‘cg_peratom’ was not declared in this scope; did you mean ‘direct_peratom’?
  203 |         cg_peratom[levels-1]->reverse_comm(this,REVERSE_AD_PERATOM);
      |         ^~~~~~~~~~
      |         direct_peratom
/home/akohlmey/compile/lammps/src/USER-OMP/msm_cg_omp.cpp:211:9: error: ‘cg_peratom’ was not declared in this scope; did you mean ‘direct_peratom’?
  211 |         cg_peratom[levels-1]->reverse_comm(this,REVERSE_AD_PERATOM);
      |         ^~~~~~~~~~
      |         direct_peratom
/home/akohlmey/compile/lammps/src/USER-OMP/msm_cg_omp.cpp:223:5: error: ‘cg’ was not declared in this scope; did you mean ‘gc’?
  223 |     cg[n]->reverse_comm(this,REVERSE_AD);
      |     ^~
      |     gc
/home/akohlmey/compile/lammps/src/USER-OMP/msm_cg_omp.cpp:228:7: error: ‘cg_peratom’ was not declared in this scope; did you mean ‘direct_peratom’?
  228 |       cg_peratom[n]->reverse_comm(this,REVERSE_AD_PERATOM);
      |       ^~~~~~~~~~
      |       direct_peratom
/home/akohlmey/compile/lammps/src/USER-OMP/msm_cg_omp.cpp:240:5: error: ‘cg_peratom_all’ was not declared in this scope
  240 |     cg_peratom_all->forward_comm(this,FORWARD_AD_PERATOM);
      |     ^~~~~~~~~~~~~~
```

File: src/KSPACE/gridcomm.cpp [hunk 1]
```diff
@@ -469,39 +510,473 @@ void GridComm::setup()
...
   }
-  nbuf *= MAX(nforward,nreverse);
-  memory->create(buf1,nbuf,"Commgrid:buf1");
-  memory->create(buf2,nbuf,"Commgrid:buf2");
+
+  nbuf2 = MAX(nbufs,nbufr);
+}
+
+/* ----------------------------------------------------------------------
+   NOTE: need to doc this header
```
Comment (stanmoore1) [inline src/KSPACE/gridcomm.cpp:747]: Need to fix this comment.

File: src/KSPACE/gridcomm.cpp [hunk 2]
```diff
@@ -469,39 +510,473 @@ void GridComm::setup()
...
-    nbuf = MAX(nbuf,swap[i].nunpack);
+    ngrid = MAX(ngrid,swap[i].npack);
+    ngrid = MAX(ngrid,swap[i].nunpack);
+  }
+
+  nbuf1 = nbuf2 = ngrid;
+}
+
+/* ----------------------------------------------------------------------
+   NOTE: need to doc this header
```
Comment (stanmoore1) [inline src/KSPACE/gridcomm.cpp:525]: Need to fix the comment

File: src/KSPACE/gridcomm.cpp [hunk 3]
```diff
@@ -469,39 +510,473 @@ void GridComm::setup()
...
+  // recurse with 2 split boxes
+  
+  if (splitflag) {
+    ghost_box_drop(newbox1,pbc);
+    ghost_box_drop(newbox2,newpbc);
+  }
+}
+
+/* ----------------------------------------------------------------------
+   NOTE: need to doc this header
```
Comment (stanmoore1) [inline src/KSPACE/gridcomm.cpp:827]: Need to fix this comment

File: src/KSPACE/gridcomm.cpp [hunk 4]
```diff
@@ -510,26 +985,114 @@ void GridComm::forward_comm(KSpace *kspace, int which)
...
+void GridComm::grow_overlap()
+{
+  maxoverlap += SWAPDELTA;
+  overlap = (Overlap *)
+    memory->srealloc(overlap,maxoverlap*sizeof(Overlap),"CommGrid:overlap");
+}
+
 /* ----------------------------------------------------------------------
    create 1d list of offsets into 3d array section (xlo:xhi,ylo:yhi,zlo:zhi)
    assume 3d array is allocated as (outxlo_max:outxhi_max,outylo_max:outyhi_max,
```
Comment (stanmoore1) [inline src/KSPACE/gridcomm.cpp:1098]: Should be full xyz hi/lo instead of out xyz hi/lo _max in this comment

**Comment by stanmoore1** [comment]
I'm working on the Kokkos version of GridComm now.

**Comment by sjplimp** [comment]
@akohlmey oops, didn't see there was an OMP/msm_cg.cpp - just pushed a new version for it

**Comment by sjplimp** [comment]
@akohlmey @stanmoore1 Re: pppm_stagger failing, the changes in that file were minor and look identical to what I did in one part of pppm.cpp.  Can you tell if this is a tiny discrepancy vs large?  Is at timestep 0?  Any suggestions on how to debug this?

**Comment by akohlmey** [comment]
> @akohlmey @stanmoore1 Re: pppm_stagger failing, the changes in that file were minor and look identical to what I did in one part of pppm.cpp. Can you tell if this is a tiny discrepancy vs large? Is at timestep 0? Any suggestions on how to debug this?

@sjplimp I just modified the peptide example to use pppm/stagger and compared an old executable with a new one. The difference appears to be due to the grid selection and the differences in the forces are substantial. From the log of the old executable:
```
PPPM initialization ...
  using 12-bit tables for long-range coulomb (src/kspace.cpp:330)
  G vector (1/distance) = 0.26787739
  grid = 12 12 12
  stencil order = 5
  estimated absolute RMS force accuracy = 0.023882441
  estimated relative force accuracy = 7.1921262e-05
  using double precision FFTW3
  3d grid and FFT values/proc = 5832 1728
```
and the new executable:
```
PPM initialization ...
  using 12-bit tables for long-range coulomb (src/kspace.cpp:330)
  G vector (1/distance) = -0.1042434
  grid = 2 2 2
  stencil order = 5
  estimated absolute RMS force accuracy = 10.533374
  estimated relative force accuracy = 0.031720943
  using double precision FFTW3
  3d grid and FFT values/proc = 512 8
```

**Comment by sjplimp** [comment]
@akohlmey thanks - that's helpful info - now I know where to look

**Comment by stanmoore1** [comment]
I started ported these changes to `PPPMKokkos`.

**Comment by sjplimp** [comment]
@akohlmey @stanmoore1 Hopefully fixed pppm/stagger.  On step 0 of peptide problem the grid is now slightly smaller (vs PPPM) and the energies slightly different.  But I'm thinking that is normal behavior.

**Comment by akohlmey** [comment]
> @akohlmey @stanmoore1 Hopefully fixed pppm/stagger. On step 0 of peptide problem the grid is now slightly smaller (vs PPPM) and the energies slightly different. But I'm thinking that is normal behavior.

Yes. It reproduces the results from the old code within machine precision:
```
$ ctest -R KSpaceStyle 
Test project /home/akohlmey/compile/lammps/build-clang
      Start 181: KSpaceStyle:ewald
 1/21 Test #181: KSpaceStyle:ewald .................   Passed    0.54 sec
      Start 182: KSpaceStyle:ewald_disp
 2/21 Test #182: KSpaceStyle:ewald_disp ............   Passed    0.56 sec
      Start 183: KSpaceStyle:ewald_nozforce
 3/21 Test #183: KSpaceStyle:ewald_nozforce ........   Passed    0.53 sec
      Start 184: KSpaceStyle:ewald_slab
 4/21 Test #184: KSpaceStyle:ewald_slab ............   Passed    0.55 sec
      Start 185: KSpaceStyle:msm
 5/21 Test #185: KSpaceStyle:msm ...................   Passed    0.83 sec
      Start 186: KSpaceStyle:msm_cg
 6/21 Test #186: KSpaceStyle:msm_cg ................   Passed    0.83 sec
      Start 187: KSpaceStyle:pppm
 7/21 Test #187: KSpaceStyle:pppm ..................   Passed    0.57 sec
      Start 188: KSpaceStyle:pppm_ad
 8/21 Test #188: KSpaceStyle:pppm_ad ...............   Passed    0.66 sec
      Start 189: KSpaceStyle:pppm_cg
 9/21 Test #189: KSpaceStyle:pppm_cg ...............   Passed    0.57 sec
      Start 190: KSpaceStyle:pppm_cg_ad
10/21 Test #190: KSpaceStyle:pppm_cg_ad ............   Passed    0.66 sec
      Start 191: KSpaceStyle:pppm_disp
11/21 Test #191: KSpaceStyle:pppm_disp .............   Passed    3.33 sec
      Start 192: KSpaceStyle:pppm_disp_ad
12/21 Test #192: KSpaceStyle:pppm_disp_ad ..........   Passed    4.65 sec
      Start 193: KSpaceStyle:pppm_disp_tip4p
13/21 Test #193: KSpaceStyle:pppm_disp_tip4p .......   Passed    2.75 sec
      Start 194: KSpaceStyle:pppm_nozforce
14/21 Test #194: KSpaceStyle:pppm_nozforce .........   Passed    0.58 sec
      Start 195: KSpaceStyle:pppm_slab
15/21 Test #195: KSpaceStyle:pppm_slab .............   Passed    0.58 sec
      Start 196: KSpaceStyle:pppm_stagger
16/21 Test #196: KSpaceStyle:pppm_stagger ..........   Passed    1.30 sec
      Start 197: KSpaceStyle:pppm_tip4p
17/21 Test #197: KSpaceStyle:pppm_tip4p ............   Passed    0.57 sec
      Start 198: KSpaceStyle:pppm_tip4p_ad
18/21 Test #198: KSpaceStyle:pppm_tip4p_ad .........   Passed    0.66 sec
      Start 199: KSpaceStyle:pppm_tip4p_nozforce
19/21 Test #199: KSpaceStyle:pppm_tip4p_nozforce ...   Passed    0.57 sec
      Start 200: KSpaceStyle:pppm_tip4p_slab
20/21 Test #200: KSpaceStyle:pppm_tip4p_slab .......   Passed    0.57 sec
      Start 201: KSpaceStyle:pppm_tri
21/21 Test #201: KSpaceStyle:pppm_tri ..............   Passed    0.58 sec

100% tests passed, 0 tests failed out of 21

Total Test time (real) =  22.45 sec
```

**Comment by stanmoore1** [comment]
Great! I should have the Kokkos version finished sometime next week.

**Comment by stanmoore1** [comment]
I have this compiling but need to test. @akohlmey is there any way to run the KSpace unit tests with Kokkos enabled?

**Comment by akohlmey** [comment]
> I have this compiling but need to test. @akohlmey is there any way to run the KSpace unit tests with Kokkos enabled?

Not directly. I'd have to hack the tester tool and disable some checks. will let you know in a while how this turns out.

**Comment by akohlmey** [comment]
> I have this compiling but need to test. @akohlmey is there any way to run the KSpace unit tests with Kokkos enabled?

It segfaults with the peptide example when using more than one MPI rank after inserting `comm_style tiled` and `balance 0.0 rcb`. Otherwise forces and energies match. You also need to make the following change:
```
  diff --git a/src/KOKKOS/pppm_kokkos.cpp b/src/KOKKOS/pppm_kokkos.cpp
  index 2167b1d8c7..ccff40ca0c 100644
  --- a/src/KOKKOS/pppm_kokkos.cpp
  +++ b/src/KOKKOS/pppm_kokkos.cpp
  @@ -191,9 +191,6 @@ void PPPMKokkos<DeviceType>::init()
                  "slab correction");
     if (domain->dimension == 2) error->all(FLERR,
                                            "Cannot use PPPM with 2d simulation");
  -  if (comm->style != 0)
  -    error->universe_all(FLERR,"PPPM can only currently be used with "
  -                        "comm_style brick");
   
     if (!atomKK->q_flag) error->all(FLERR,"Kspace style requires atom attribute q");
   
```

**Comment by stanmoore1** [comment]
OK thanks, I'll look into the segfault

**Comment by stanmoore1** [comment]
This is now working for both Kokkos OpenMP and CUDA backends.

**Comment by akohlmey** [comment]
@sjplimp this is reassigned to you for final review.

**Comment by stanmoore1** [comment]
I'm done with my changes (unless more issues are discovered). Do we want to do a  full regression before merging?

**Comment by sjplimp** [comment]
yes, good idea to do a full regression test.  However, hold off a bit.  I want to make one other change to the FFTs that I believe will have no effect on results.

**Comment by akohlmey** [comment]
> yes, good idea to do a full regression test. However, hold off a bit. I want to make one other change to the FFTs that I believe will have no effect on results.

Another thing that needs to be done is to update the PPPM implementation in the GPU package. Cross-compilation of the windows package is failing because of that.
```
/home/jenkins/workspace/dev/pull_requests/compilation_tests/lammps/src/GPU/pppm_gpu.cpp: In member function 'virtual void LAMMPS_NS::PPPMGPU::unpack_reverse_grid(int, void*, int, int*)':
/home/jenkins/workspace/dev/pull_requests/compilation_tests/lammps/src/GPU/pppm_gpu.cpp:658:15: error: declaration of 'FFT_SCALAR* buf' shadows a parameter
  658 |   FFT_SCALAR *buf = (FFT_SCALAR *) vbuf;
      |               ^~~
/home/jenkins/workspace/dev/pull_requests/compilation_tests/lammps/src/GPU/pppm_gpu.cpp:658:36: error: 'vbuf' was not declared in this scope; did you mean 'buf'?
  658 |   FFT_SCALAR *buf = (FFT_SCALAR *) vbuf;
      |                                    ^~~~
      |                                    buf
/home/jenkins/workspace/dev/pull_requests/compilation_tests/lammps/src/GPU/pppm_gpu.cpp:656:51: warning: unused parameter 'buf' [-Wunused-parameter]
  656 | void PPPMGPU::unpack_reverse_grid(int flag, void *buf, int nlist, int *list)
```

**Comment by stanmoore1** [comment]
>Another thing that needs to be done is to update the PPPM implementation in the GPU package

I fixed that typo in pppm_gpu.cpp.

**Comment by stanmoore1** [comment]
Kokkos regression tests passed.

**Comment by sjplimp** [comment]
@akohlmey @stanmoore1 Ok - my changes are done - please regress away ...

**Comment by akohlmey** [comment]
> @akohlmey @stanmoore1 Ok - my changes are done - please regress away ...

@sjplimp FYI, you have introduced a **lot** of tabs and trailing whitespace. Will have to use the extra strength cleaning brush to get rid of those... :-(

File: src/KSPACE/fft3d.cpp [hunk 5]
```diff
@@ -103,18 +103,18 @@ void fft_3d(FFT_DATA *in, FFT_DATA *out, int flag, struct fft_plan_3d *plan)
   length = plan->length1;
 
 #if defined(FFT_MKL)
-  if (flag == -1)
+  if (flag == 1)
```
Comment (akohlmey) [inline src/KSPACE/fft3d.cpp]: @sjplimp This change and the following below break all PPPM styles. Am reverting it.

Comment (sjplimp) [reply src/KSPACE/fft3d.cpp]: ok - will take a look - thanks

**Comment by ndtrung81** [comment]
@akohlmey I suppose everything is good, and there's no need for my review at this point.

**Comment by akohlmey** [comment]
> @akohlmey I suppose everything is good, and there's no need for my review at this point.

yes, your review was auto-requested because you are listed as a code owner for one of the affected files in `.github/CODEOWNERS`, probably pppm/gpu.

**Comment by akohlmey** [comment]
@sjplimp i've resolved the merge conflict and added a couple of unit tests for kspace with comm_style tiled (won't really test the communication patterns, but will help us detect when changes have side effects). I will now trigger the full regression tests one more time. Anything else that needs to be done before this is ready to merge?

**Comment by stanmoore1** [comment]
We figured out how to get correct forces when flipping FFT forward/reverse (that @sjplimp tried to add). Do we just want to make that a new PR and add that later?

**Comment by akohlmey** [comment]
> We figured out how to get correct forces when flipping FFT forward/reverse (that @sjplimp tried to add). Do we just want to make that a new PR and add that later?

Yes. You could make a feature branch from this branch and then submit this as PR. If you compile with cmake and enable testing, you can then test the kspace styles selectively with `ctest -R KSpace`.

**Comment by sjplimp** [comment]
@akohlmey @stanmoore1 I'm a bit lost in the git lingo.  If you're saying you want to (a) drop any FFT changes from this PR and test/merge it now, and then (b) do another PR later to make the FFT and PPPM changes, then that's fine by me.

**Comment by akohlmey** [comment]
> @akohlmey @stanmoore1 I'm a bit lost in the git lingo. If you're saying you want to (a) drop any FFT changes from this PR and test/merge it now, and then (b) do another PR later to make the FFT and PPPM changes, then that's fine by me.

yes. All I was saying is that it is possible to create the branch with `git checkout -b fft-feature gridcomm-tiled`, so that you have all the changes from here already included instead of starting from master.

I am currently trying to identify some issue that was flagged by one of the load balancing regression tests but cannot (yet) reproduce it.

**Pull request merged**